protocols/v2/roles-logic-sv2/src/utils.rs

```diff
@@ -276,48 +280,94 @@ impl TryFrom<CoinbaseOutput> for Script {
 /// [3] https://en.wikipedia.org/wiki/Negative_hypergeometric_distribution
 /// bdiff: 0x00000000ffff0000000000000000000000000000000000000000000000000000
 /// https://en.bitcoin.it/wiki/Difficulty#How_soon_might_I_expect_to_generate_a_block.3F
-pub fn hash_rate_to_target(h: f32, share_per_min: f32) -> U256<'static> {
+
+pub fn hash_rate_to_target(
+    hashrate: f32,
+    share_per_min: f32,
+) -> Result<U256<'static>, crate::Error> {
+    // checks that we are not dividing by zero
+    if share_per_min == 0.0 {
+        return Err(Error::ImpossibleToGetTarget);
```

Comment:
Maybe is better to have more granular error like ZeroNotAllowed

Reply:
This error that you suggest it is the same for the functions hash_rate_from_target and hash_rate_to_target. For having the same granularity as the PR it should be differnetiated like this
ZeroNotAllowedForCalculatingTarget
ZeroNotAllowedForCalculatingHashrate
which is the same as it is now, but with different names.

Reply:
you can do either like that or TargetErro(ZeroInput) and HashRateErro(ZeroInput) or just ZeroInputError but would be nice to see in the error name that the issue is that we are using a vlaue that is 0. That cause when you call the function you immediately know that you are not allowed to pass values that are 0 and you act consequentially.  With the other error name you don't know why tha function that you are calling could fail and you are forced to go and look at the implementation to understand why the function can fail.

Reply:
Okok understood.
Do you like this?
 TargetError(DivisionByZero) 
 HashrateError(DivisionByZero)

Reply:
ok

Reply:
but it happens only when the input is 0 so just say in the error that 0 input are not allowed is a lot more lil;ey  that the caller will handle the error in the correct way. You know that you have the error only in a very aprticular case just say it in the error

Reply:
The inputs can be bad in several ways. If they are negative for example this is bad.
I handled also this case, just a sec that I finish to apply your suggestions!